rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+relatively simple operators (e.g. `SELECT`, `WHERE`,  `JOIN`) is transformed
+into a much more specific, optimized query plan. Predicates from a `WHERE`
+clause are pushed down, the theoretical `JOIN` becomes a concrete hash join,
+merge join, or nested loop, etc, etc. Vector will not need the same level of
+sophistication as PostgreSQL's query planner and optimizer, but the idea of
+processing user input into an intelligent execution plan is the same.
+
+Concretely, this will affect both the internal design of our components as well
+as the way that we combine them into topologies.
+
+### Component Design
+
+Our current component designs are extremely bare-bones and reflect the way that
+we directly turn them into physical tasks. To be able to plan more
+intelligently, we'll need a larger, more precise vocabulary.
+
+#### Sources
+
+Of all the component types in Vector, sources are the least effected by this
+proposal. They are currently modeled as essentially a bare `Task` that is given
+an output channel. This reflects the wide variety of source implementations,
+from HTTP servers to file tailing to stdin.
+
+The only thing we seek to change with this proposal is the implementation of the
+output channel. Instead of a literal concrete channel sender, sources will
+receive an implementation of a new `Push` trait (similar to the one in the
+`[timely_communication`
+crate](https://docs.rs/timely_communication/0.11.1/timely_communication/trait.Push.html)).
+The API will be roughly the same as before, but give us more flexibility in
+implementation behind the scenes.
+
+#### Transforms
+
+Transforms are currently represented by a single trait with three methods
+(`transform`, `transform_into`, and `transform_stream`). These methods roughly
+represent different levels of capabilities and a given implementation should
+really only use one of them.
+
+With this change, we will break up the singular `Transform` trait into a number
+of different traits that directly represent these various capabilities (as well
+as some new ones). Initially, the biggest differentiators will be:
+
+1. Persistent state / ordering
+2. Ability to produce events independently of incoming events
+
+The first is currently implicit and we assume all transforms could be accumulate
+state and require strict ordering for correct execution. Since we currently use
+literal serial execution, this is supported by default.
+
+The second was the original impetus for `transform_stream`. This determines
+whether a transform can be a simple function like `map` or `filter`, or if it
+needs to be its own task that can be scheduled independently. Currently all
+transforms are run as a task either way, and this is just  a matter of
+implementation convenience.
+
+Our model of transforms will adapt into something like the following pseudocode:
+
+```rust
+trait TransformConfig {
+    async fn build(&self) -> Transform
+}
+
+enum Transform {
+    Fn(TransformFn),
+    StatefulFn(TransformFn),
```

Comment:
Do you think it makes sense to have a distinction between like `LocalStateFn` and `SharedStateFn`? `LocalStateFn` could probably be optimized similarly to `Fn`.

Reply:
If `LocalStateFn` means state is purely internal and not required for correctness (e.g. reusing capturer allocations in `regex_parser`), then I'd consider that `Fn`. Otherwise, I think the distinction is less about local vs remote and more about whether state and ordering affects output.